addon/components/connect.js

```diff
@@ -1,98 +1,113 @@
 import Ember from 'ember';
 
 const {
+  assert,
   computed,
   defineProperty,
+  getProperties,
   inject: { service },
-  isEmpty,
   run
 } = Ember;
 
-export default (stateToComputed, dispatchToActions=() => ({})) => {
+/**
+  Returns a list of keys that have different values between the two objects.
+
+  @method changedKeys
+  @return {Array} keys that have changed
+  @private
+*/
+function changedKeys(state, newState) {
+  return Object.keys(state).filter(key => {
+    return state[key] !== newState[key];
+  });
+}
+
+/**
+  Creates a read-only computed property for accessing redux state.
+
+  @method computedReduxProperty
+  @return {Function} an Ember computed property
+  @private
+*/
+function computedReduxProperty(key, getState) {
+  return computed({
+    get: () => getState()[key],
+    set: () => assert(`Cannot set redux property "${key}".`)
+  });
+}
+
+/**
+  Return an object of attrs passed to this Component.
+
+  `Component.attrs` is an object that can look like this:
+
+    {
+      myAttr: {
+        value: 'myValue'
+      }
+    }
 
-  if (!stateToComputed) {
-    stateToComputed = () => ({});
-  }
+  Ember provides that a `get` will return the value:
 
-  return Component => {
+    this.get('myAttr') === 'myValue'
 
-    return Component.extend({
+  @method getAttrs
+  @return {Object} an object of key, value for each attr
+  @private
+*/
+function getAttrs(context) {
+  const attrKeys = Object.keys(context.attrs || {});
+  return getProperties(context, attrKeys);
+}
 
+export default (stateToComputed, dispatchToActions) => {
+  return Component => {
+    return Component.extend({
       redux: service(),
 
       init() {
         const redux = this.get('redux');
 
-        let props = stateToComputed.call(this, redux.getState(), this.getAttrs());
+        if (stateToComputed) {
+          const getState = () => stateToComputed.call(this, redux.getState(), getAttrs(this));
 
-        Object.keys(props).forEach(name => {
-          defineProperty(this, name, computed(() =>
-            stateToComputed.call(this, redux.getState(), this.getAttrs())[name]
-          ).property().readOnly());
-        });
+          let state = getState();
```

Comment:
This is the bulk of the value add here correct? We cut down on the work required because you call `getState` just 1x now. The tradeoff here is that we are storing state locally in the component itself - any challenges/ or issues with this? I was afraid of the idea conceptually because although we called `getState` 2x we didn't need to worry about state synchronization (as each call for the state backing a CP would call out for it on demand)

Just curious what we are trading here if anything